optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/drools/KieBaseExtractor.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2021 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaplanner.core.impl.score.director.drools;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import org.drools.core.io.impl.ClassPathResource;
+import org.drools.core.io.impl.FileSystemResource;
+import org.drools.modelcompiler.ExecutableModelProject;
+import org.kie.api.KieBase;
+import org.kie.api.conf.KieBaseMutabilityOption;
+import org.kie.internal.builder.conf.PropertySpecificOption;
+import org.kie.internal.utils.KieHelper;
+import org.optaplanner.core.config.score.director.ScoreDirectorFactoryConfig;
+import org.optaplanner.core.config.util.ConfigUtils;
+
+public class KieBaseExtractor {
+
+    private static Map<String, KieBase> scoreDrlListToKieBaseMap = new HashMap<>();
```

Comment:
This is a memory leak, and a pretty severe one. This must not be `static`.

Reply:
Would have a static KieRuntimeBuilder (which is null unless Quarkus sets it) result in a memory leak?

(it would look like this (the map would be removed):
```
if (kieRuntimeBuilder != null) {
    return kieRuntimeBuilder.getKieBase();
} else {
    // code to construct KIE base w/o storing in a map
}
```

Reply:
If you store KieBases somewhere and never remove them for the runtime of the JVM, you are leaking KIeBases. There is no way around that.

Why do you need to store them? I'm assuming it's a cache. In that case, I suggest you figure out a way to make the cache local to something that eventually gets garbage-collected.

Reply:
To pass an injected `KieBaseBuilder` to the `KieBaseExtractor` (which is used in `ScoreDirectorFactoryFactory`); 

A Gizmo class is generated that inject the KieBaseBuilder and pass it to the `KieBaseExtractor` (currently it passes the `KieBase`, but it can pass the `KieBaseBuilder` instead.)

As for why, we are using the generated KieBase from the Kogito plugin (we aren't creating a new one, but reusing an existing one).

Reply:
Let's discuss in person tomorrow. There has to be a way to get this on-demand, locally, as opposed to storing it somewhere for the lifetime of the JVM. Please schedule a 30 minute call, and if you could start with a quick walk-through of the relevant code, that'd be useful.